[Jul 30] Running benchmarks on GitHub Actions (#217)
Quansight-Labs/quansight-labs-site: posts/2021/08/github-actions-benchmarks.md

<!--
.. title: Is GitHub Actions suitable for running benchmarks?
.. slug: github-actions-benchmarks
.. date: 2021-08-30 00:01 UTC
.. tags: github-actions, Open-Source, continuous-integration, performance
.. category:
.. link:
.. description:
.. type: markdown
.. author: Jaime Rodríguez-Guerra
.. previewimage: /images/2021/08/github-actions-benchmark.png
.. image:: /images/2021/08/github-actions-benchmark.png
   :alt: Reliability of benchmarks in GitHub Actions
-->

> ![Reliability of benchmarks in GitHub Actions](/images/2021/08/github-actions-benchmark.png)

Reviewer: It is a bit ambiguous at this point (before reading the below text) what the individual dots here represent, but maybe it is still best to keep this colorful figure up front to grab attention?

Author: Yes, I agree. I only added it here for eye candy, but if needed we can think of a different banner?

the last release? We do not need absolute measurements, just a way to _compare_ two commits in
a reliable way. This is exactly what we meant by _relative benchmarking_.

The idea is to run the benchmarking suite in the same CI job, twice, once for each commit. This
way, we guarantee we are using the exact same machine and configuration for both. By doing this, we
are essentially canceling out the background noise and can estimate the performance _ratio_ of
the two commits.

There are some gotchas, though. Benchmark suites can take long to run, and running them twice
makes it even worse. This opens an even larger window for other CI jobs to pollute our
measurements with resource-intensive tasks that happen to be running at the same time.
To minimize these effects, benchmarking tools run the tests several times in different
schedules and apply some statistics to correct for the inevitable deviations. Or, at least,
_try_.

# Relative performance measurements with Airspeed Velocity

`scikit-image`, the project that commissioned this task, uses
[Airspeed Velocity](https://asv.readthedocs.io/), or `asv`, for their benchmark tests.

> `asv`'s main feature is being able to track performance measurements overtime in a JSON database and generate

Reviewer: ```suggestion
`asv`'s main feature is being able to track performance measurements over time in a JSON database and generate
```

involves compiled libraries (as is the case with `scikit-image`), this can be a lengthy
process!
2. The default configuration will run the benchmark in two interleaved passes, with several
repeats each. This is done to account for the unavoidable deviations from ideality caused
from co-running processes, as mentioned above.
3. The statistics for each commit are gathered and a report table is presented. The ratio
of each test is computed and, if it is greater than a certain threshold (`1.2` by default)
an error is emitted.

> <sup>†</sup> `asv` supports the notion of a configuration matrix, so you can test your code under
different environments; e.g. NumPy versions, Python interpreters, etc.

For the benchmark suite of `scikit-image` as of June/July 2021, this ends up taking up to two hours.
This raises two questions we will answer in the following sections:

1. Although we are trying hard to reduce measurement errors, is that enough? Are these measurements reliable?
2. Two hours might be too long. Are there any settings we can tune to reduce the runtime
without reducing the accuracy of the measurements?

# Are CI services reliable enough for benchmarking?

> We saw above that CI services are not designed for this kind of tasks, but with some workarounds

Reviewer: ```suggestion
We saw above that CI services are not designed for this kind of task, but with some workarounds
```

they might be good enough. However, how can be sure? As data-driven scientists, we say let's run
an experiment!

## The setup

> Our experiment is data-drive, so we need to generate some data first. This is our strategy:

Reviewer 2: data-driven?

Author: ```suggestion
Our experiment is data-driven, so we need to generate some data first. This is our strategy:
```

some kind of error is expected. We want to see if it stays reliably under an acceptable threshold.
* We will implement a GitHub Actions workflow that will run every six hours (four times a day) and
collect results for a week or more. This will help us account for two things:
    * Accumulate sufficient data points to answer our question.
    * Account for time factors like different days of the week (e.g. weekday vs weekend),
      or the time of the day (3am vs 3pm).
* The workflow will upload the benchmark results as artifacts we can download and process locally
  with a Jupyter Notebook.

> Check the [GitHub Actions workflow in this fork](https://github.com/jaimergp/scikit-image/blob/main/.github/workflows/benchmarks-cron.yml)!

## The results

To measure the stability of the different benchmark runs we will be looking at the performance
ratio of the measurements between the two commits. Since they are identical in terms of tested code,
ideally they should be all `1.0`. We know this will not happen, but maybe the errors are not that big
and stay like that regardless the time of the day or the day of the week.

> After some days collecting data points, these are the results:

Reviewer: perhaps specify the actual number of days here

After collecting data for XX days, ...

![Reliability of benchmarks in GitHub Actions](/images/2021/08/github-actions-benchmark.png)

* Average time taken: 1h55min
* Minimum and maximum ratios observed: 0.51, 1.36
* Mean and standard deviation of ratios: 1.00, 0.05
* Proportion of false positives: 4/108 = 3.7%

In the X axis you can see the different runs, sorted by date and time. Days of the week are grouped
with colored patches for easier visual analysis. In the Y axis, we plot the performance ratio. Each
of those vertical clouds include 75 points, one per benchmark test. Ideally, they should all fall at
`y=1`. Of course, not all of them are there, but a surprising amount of them do!

But! We do see some bigger deviations at certain times! How is that possible? Well, that's our error!
These outliers are the ones that would concern us because they can be interpreted as **false positives**.
In other words: `asv` would report a performance regression, when in fact there's none. However,

> in the observed measurements, the outliers were always within `y ∈ (0.66, 1.5)`.

Reviewer: This line states 0.66, but the "minimum" that was stated a paragraph or two before was 0.51. seems like a contradiction?

Author: You are right, let me look into this!

* Average time taken: 1h39min
* Minimum and maximum ratios observed: 0.43, 1.5
* Mean and standard deviation of ratios: 0.99, 0.07
* Proportion of false positives: 6/66 = 9.99%

## Single-pass with `processes=1`

In this configuration, we expect a drastic 50% running time reduction, since we will only do one pass
per commit, instead of two. However, the accuracy loss might be too dramatic... Let's see!

![Reliability of benchmarks in GitHub Actions, single-process](/images/2021/08/github-actions-benchmark-single-process.png)

* Average time taken: 1h7min
* Minimum and maximum ratios observed: 0.51, 2.76
* Mean and standard deviation of ratios: 1.01, 0.07
* Proportion of false positives: 8/64 = 12.5%

At first sight, those clouds look very spread! The number of false positives is also larger.

## Summary of the strategies

Let's take a look at the three strategies now. The main columns are **runtime** and **%FP**
(percentage of false positives). We want the smallest %FP at the shortest runtime.

| Strategy        | Runtime  &nbsp;| %FP  | Min  | Max  | Mean &nbsp;| Std  |
|-----------------|---------|-------|------|------|------|------|

> | Default         | 1h55    | 3.7%  | 0.51 &nbsp;| 1.36 | 1.00 | 0.05 |

Author: ```suggestion
| Default         | 1h55    | 3.7  | 0.51 &nbsp;| 1.36 | 1.00 | 0.05 |
```

> | No interleaving &nbsp;| 1h39    | 9.99% &nbsp;| 0.43 | 1.50 | 0.99 | 0.07 |

Reviewer: ```suggestion
| No interleaving &nbsp;| 1h39    | 9.99 &nbsp;| 0.43 | 1.50 | 0.99 | 0.07 |
```

(The other rows don't have % here and it is stated in the header row)

Author: I think they do have the `%`, but yes, we can remove it if it's in the header!

> | Single pass     | 1h07    | 12.5% | 0.51 | 2.76 &nbsp;| 1.01 | 0.07 |

Author: ```suggestion
| Single pass     | 1h07    | 12.5 | 0.51 | 2.76 &nbsp;| 1.01 | 0.07 |
```

# TLDR

We have seen that GitHub Actions is indeed good enough to perform relative benchmarking in a reliable way
as long as several passes are averaged together (default behaviour in `asv continuous`). This takes a bit
more time but you can speed it up a bit with `mamba` and `ccache` for compiled libraries. Even in that case,
it is probably overkill to run it for every push event, so we are using the `pull_request.labeled` trigger
to let the maintainers decide when to do it on demand.

---

# Useful references

* [scikit-image/scikit-image #5424](https://github.com/scikit-image/scikit-image/pull/5424): The PR where
  all this was implemented.
* [asv.readthedocs.io](https://asv.readthedocs.io): The official documentation for `asv`.
* [Building an Open Source, Continuous Benchmark System](https://wolfv.medium.com/building-an-open-source-continuous-benchmark-system-717839093962)
* [Conbench: Language-independent Continuous Benchmarking Tool, by Ursa Labs](https://ursalabs.org/blog/announcing-conbench/)
* [PyPy Speed](https://speed.pypy.org/)
* [Pandas Speed](http://pandas.pydata.org/speed/pandas/)

# Acknowledgements

> Thanks Gregory Lee, Stefan van der Walt and Juan Nunez-Iglesias for their enthusiastic and useful feedback!

Reviewer: ```suggestion
Thanks Gregory Lee, Stéfan van der Walt and Juan Nunez-Iglesias for their enthusiastic and useful feedback!
```

> The plots look that pretty thanks to comments provided by John Lee.

Reviewer: please add a funding acknowledgement similar to:

This work was funded by the Chan-Zuckerberg Institute (CZI) as part of an [Essential Open Source Software for Science grant](https://chanzuckerberg.com/eoss/proposals/gpu-acceleration-rapid-releases-and-biomedical-examples-for-scikit-image/).

The help message for this flag says:

    > Interleave benchmarks with multiple processes across commits. This can avoid measurement biases
    > from commit ordering, can take longer.

    How much longer? Does it help keep error under control? We should measure that.

- By default, all tests are run several times, with different schedules. There are two
[benchmark attributes](https://asv.readthedocs.io/en/stable/benchmarks.html#timing-benchmarks)
that govern these settings: `processes` and `repeat`. `processes` defaults to `2`, which means that
the full suite will be run twice per commit. If we only do one pass (`processes=1`), we will reduce
the running time in half, but will we lose too much accuracy?

To answer both questions, we added more entries to the data collection workflow shown above by
[parameterizing the `asv` command-line options](https://github.com/jaimergp/scikit-image/blob/e561996/.github/workflows/benchmarks-cron.yml#L14-L23).

These are the results!

## No process interleaving

Disabling process interleaving should be faster and maybe the accuracy loss is not that bad. But...
how bad? Here are the results:

> ![Reliability of benchmarks in GitHub Actions, no interleaving. This 2D plots shows a 16-day timeseries in the X axis.

Reviewer 3: Because you are a) comparing similar plots and (correctly) repeating alt text and b) relying heavily on the image to tell the story instead of summarizing it elsewhere, I recommend moving the intended takeaway of the plot to earlier in the alt text like:

```
Reliability of benchmarks in GitHub Actions, no interleaving. One particular 
outlier happened on the first Saturday with half the cloud below 0.75 performance. 
This 2D plots shows a 16-day timeseries in the X axis.
Each data point in the X axis corresponds to a cloud of 75 measurements (one per benchmark test).
The y-axis spread of each cloud corresponds to the performance ratio. Ideal measurements would have
a performance ratio of 1.0, since both runs returned the exact same performance. In practice this
does not happen. With no interleaving, the vertical spread is more evident, with several clouds
spreading beyond the desired interval.
```

While this doesn't quite follow the recommendations I posted, I think it makes more sense for screen reader users (and maybe people with cognitive disabilities) to be able to get the info they need most at the start and feel like they can skip the parts that are repeated later if they want.

  Each data point in the X axis corresponds to a cloud of 75 measurements (one per benchmark test).
  The y-axis spread of each cloud corresponds to the performance ratio. Ideal measurements would have
  a performance ratio of 1.0, since both runs returned the exact same performance. In practice this
  does not happen. With no interleaving, the vertical spread is more evident, with several clouds
  spreading beyond the desired interval. One particular outlier happened on the first Saturday,
  with half the cloud below 0.75.](/images/2021/08/github-actions-benchmark-no-interleaving.png)

* Average time taken: 1h39min
* Minimum and maximum ratios observed: 0.43, 1.5
* Mean and standard deviation of ratios: 0.99, 0.07
* Proportion of false positives: 6/66 = 9.99%

## Single-pass with `processes=1`

In this configuration, we expect a drastic 50% running time reduction, since we will only do one pass
per commit, instead of two. However, the accuracy loss might be too dramatic... Let's see!

> ![Reliability of benchmarks in GitHub Actions, single pass. This 2D plots shows a 16-day timeseries in the X axis.

Reviewer 3: For the same reason as mentioned in the third alt text, I'd recommend moving the main takeaway earlier in the alt text like 

```
Reliability of benchmarks in GitHub Actions, single pass. With a single process, 
the y-spread is significantly wider. Some data points can
be observed even beyond a ratio of 2.5 (it should be 1.0!), and the overall visual spread is
larger; i.e. the clouds are taller than in the ideal case of having a small, contained cloud at 1.0.
This 2D plots shows a 16-day timeseries in the X axis.
Each data point in the X axis corresponds to a cloud of 75 measurements (one per benchmark test).
The y-axis spread of each cloud corresponds to the performance ratio. Ideal measurements would have
a performance ratio of 1.0, since both runs returned the exact same performance. In practice this
does not happen. 

```
